### PR TITLE
Movie Ticket Application - User Signup and Signin Process, Ticket Add and Book Based on Difference Role Host & Watcher

### PR DESCRIPTION
**Pull request by ghost**
Quick Ticket is a python based Movie Ticket booking application, implemented using classes and object with json & file system operations.

Features:

- User Signup and Signin Implementation
- Seperate Role for Host and Watcher
- Host allowed to add movies and watch movies.
- Watcher allowed to watch movie only
- Real-time update of Json File
- Date slots are based on current date
- Time slots are based on current time

[Screencast from 08-07-24 12:19:03 PM IST.webm](https://github.com/mohit-trootech/Quick-Ticket/assets/172255945/999e3c5f-e341-417e-a9a2-6fb0e1c18c7a)

### CODE REVIEW

File: constant.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+DATA_OPTION = dict(enumerate(["USER_JSON", "MOVIE_JSON"], 1))
+USER_JSON = DATA_OPTION.get(1)
+MOVIE_JSON = DATA_OPTION.get(2)
+
+
+# User Role
+ROLE = tuple(enumerate(["host", "watcher"], 1))
+ROLE_HOST = ROLE[0][1]
+ROLE_WATCHER = ROLE[1][1]
+
```
Comment (Dhruvi-Makwana) [inline constant.py]: why take role like this ?
you can take dict as well.

File: exceptions.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+class DataNotFound(BaseException):
+    """custom exception for tickets unavailable"""
+
+    def __init__(self, msg):
+        super(DataNotFound, self).__init__(msg)
+
+
+class UnderAge(BaseException):
+    """custom exception for tickets unavailable"""
```
Comment (Dhruvi-Makwana) [inline exceptions.py:33]: you can changed name to "ineligible"

Comment (ghost) [reply exceptions.py:33]: Updated

File: constant.py (hunk 1 above)
Comment (ghost) [reply constant.py]: Updated

**Review by Dhruvi-Makwana — APPROVED** [review]

File: .gitignore [hunk 3]
```diff
@@ -1,2 +1,4 @@
 __pycache__
```
Comment (yatishTrootech) [inline .gitignore:1]: branch name should not be `working`, it should be related to functionality name

File: constant.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+TRANSACTION_FILENAME_FORMAT = "{username}_transactions.json"
+DATA_JSON_PATH = os.path.join(BASE_JSON_DIR, "data.json")
+DATA_JSON_FORMAT = {"USER_JSON": {}, "MOVIE_JSON": {}}
+DATA_OPTION = dict(enumerate(["USER_JSON", "MOVIE_JSON"], 1))
+USER_JSON = DATA_OPTION.get(1)
+MOVIE_JSON = DATA_OPTION.get(2)
+
+
+# User Role
+ROLE = dict(enumerate(["host", "watcher"], 1))
```
Comment (yatishTrootech) [inline constant.py]: that should be Enum

File: constant.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+"""Quick-Cart constant File to Store Constant Variable & Messages"""
+
+import os
+
+BASE_JSON_DIR = os.path.join(os.getcwd(), "quick_tickets_json")
+TRANSACTION_FILENAME_FORMAT = "{username}_transactions.json"
+DATA_JSON_PATH = os.path.join(BASE_JSON_DIR, "data.json")
+DATA_JSON_FORMAT = {"USER_JSON": {}, "MOVIE_JSON": {}}
+DATA_OPTION = dict(enumerate(["USER_JSON", "MOVIE_JSON"], 1))
```
Comment (yatishTrootech) [inline constant.py]: this is fine but we can use Enum as well

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+    if available_ticket == 0:
+        print("Houseful")
+        exit()
+    tries = 3
+    while True:
+        ticket_quantity = user_input_int_type("Select Ticket Quantity: ")
+        if ticket_quantity <= available_ticket:
+            return ticket_quantity
+        elif tries == 0:
+            raise UnavailableTickets("Try Booking the Tickets the are Available")
```
Comment (yatishTrootech) [inline utils.py]: Why this message is not in constants ?

File: exceptions.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class DateValidationError(BaseException):
+    """custom exception for tickets unavailable"""
+
+    def __init__(self, msg):
+        super(DateValidationError, self).__init__(msg)
+
+
+if __name__ == "__main__":
```
Comment (yatishTrootech) [inline exceptions.py]: why do we need this  __name__ == "__main__" ?

File: main.py
Comment (yatishTrootech) [inline main.py:1]: file should have meaning full name

File: main.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    user_input_int_type,
+    create_required_json_files,
+    create_user_transactions_json_file,
+)
+
+
+def main() -> None:
+    """
+    main Driver function for Step by Step Process
+    @return:None
```
Comment (yatishTrootech) [inline main.py]: if you are defining output type `-> None` then no need to define `@return:None`, same for parameters

File: utils.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    write_json(DATA_JSON_PATH, data)
+
+
+def create_required_json_files() -> None:
+    """
+    function to create requires json files
+    @return: None
+    """
+    if check_filepath_exists(DATA_JSON_PATH):
+        create_directory_if_not_exist("quick_tickets_json")
```
Comment (yatishTrootech) [inline utils.py]: why this string is static

File: main.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    create_required_json_files()
+    username = user_input("Please Enter Username: ")
```
Comment (yatishTrootech) [inline main.py]: again static string

File: users.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    def sign_up(self):
+        """
+        method to implement signup process
+        @return:
+        """
+        custom_print_msg("Please Enter User Information")
+        password = password_hash(input("Enter Your Password: "))
+        role = user_input_int_type(f"Please Select User Role {ROLE}: ")
+        if int(role) == 1:
```
Comment (yatishTrootech) [inline users.py]: `
if role not in [ROLE_HOST, ROLE_WATCHER]:
     raise
role = int(user_input_int_type(f"Please Select User Role {ROLE}: "))
no need if else conditions

File: users.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            role = ROLE_HOST
+        else:
+            role = ROLE_WATCHER
+        name = user_input("Enter Name: ")
+        age = user_input_int_type("Enter Age: ")
+        if age < 18:
+            raise UnderAge("You are Under Age to Book Tickets Required 18 plus")
```
Comment (yatishTrootech) [inline users.py]: Everywhere if you used static messages then replace them with constants |?

File: users.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        email = user_input("Enter Email: ")
+        status = self.insert_signup_information(password, name, int(age), email, role)
```
Comment (yatishTrootech) [inline users.py]: can't you send kwargs ?, in future if I add or remove the fields then this functions args will get updated

File: users.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        if status:
```
Comment (yatishTrootech) [inline users.py]: `self.insert_signup_information` this will never return False then why `if status` condition ?

File: users.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        method to check whether the user Exist
+        @return: None
+        """
+        data = read_json(DATA_JSON_PATH)
+        if not data.get(USER_JSON).get(self.username):
+            custom_print_msg(USER_NOT_EXIST)
+            self.sign_up()
+            return
+        else:
```
Comment (yatishTrootech) [inline users.py]: No need to write else here coz above returns in if condition

File: main.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    further_progress = user_input("Press C/c to Cancel\nElse Press Enter to Continue: ")
+    if further_progress == "c":
+        return
+    service = user_input_int_type(
+        "Please Select Service\n"
+        "1. Book Movie\n"
+        "2. Add Movie\n"
+        "Select from the Given Options in int Format: "
+    )
+    if service == 1:
```
Comment (yatishTrootech) [inline main.py]: create Enum for this 
`
class ServiceEnum(Enum):
         LIST_MOVIE = 1
         ADD_MOVIE = 2
`

then use ServiceEnum.LIST_MOVIE

File: utils.py [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+    @return:list
+    """
+    upcoming_movie = []
+    past_movies = []
+    for key, value in movie_date.items():
+        if str(date.today()) <= value.get("end_date"):
+            upcoming_movie.append(key)
+        else:
+            past_movies.append(key)
+    return [upcoming_movie, past_movies]
```
Comment (yatishTrootech) [inline utils.py]: `return upcoming_movie, past_movies` is the same thing it will return tuple

File: utils.py [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+
+def get_available_date_slots(date_slots: dict) -> dict:
+    """
+    function to return date slots counter based on current date
+    @param date_slots: dict
+    @return: dict
+    """
+    counter = 1
+    date_slot_counter = {}
+    for key in date_slots.keys():
```
Comment (yatishTrootech) [inline utils.py]: key what does it mean key is predefined keyword in python , It should be meaningful name

File: theater.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+        @return: None
+        """
+        self.movies_data = read_json(DATA_JSON_PATH).get(MOVIE_JSON)
+        date_specified_movies_list = generate_movie_list_based_on_date(self.movies_data)
+        option = user_input_int_type(
+            "Please Choose\n1. Upcoming Movies\n2. Past Movies: "
+        )
+        upcoming_movies_counter = dict(enumerate(date_specified_movies_list[0], 1))
+        past_movie_counter = dict(enumerate(date_specified_movies_list[1], 1))
+        if option == 1:
```
Comment (yatishTrootech) [inline theater.py]: static options

**Review by yatishTrootech — CHANGES_REQUESTED** [review]

**Review by yatishTrootech — APPROVED** [review]

**Pull request merged**